feat: adds option to increase rover's client timeout

EverlastingBugstopper opened the pull request:
Fixes #792 by providing an option for configuring Rover's HTTP client timeout in seconds, and adding a new error code `E031` for when HTTP(S) operations time out.

```console
$ rover config whoami --client-timeout 0
Checking identity of your API key against the registry.
error[E031]: error sending request for url (https://graphql.api.apollographql.com/api/graphql): operation timed out

Caused by:
    operation timed out
              You can try increasing the timeout value by passing a higher value to the --client-timeout option.
```

---

From [the docs]() attached to this PR (I'm not sure why Netlify didn't make a preview deploy... maybe because I didn't branch off of `main`?)

---

#### Increase the client timeout

Sometimes, you may want to make queries for large amounts of data that may take some extra time for the Studio API to process.

Rover supports increasing the client timeout in cases like this with the `--client-timeout` option, which takes a number of seconds as a parameter. By default, Rover will time out requests after 30 seconds.

EverlastingBugstopper committed:
feat: adds options to bypass SSL validation

EverlastingBugstopper committed:
feat: adds option to increase rover's client timeout

EverlastingBugstopper committed:
docs: document --client-timeout

EverlastingBugstopper committed:
s/flag/option

EverlastingBugstopper committed:
docs: include E031

EverlastingBugstopper committed:
s/SSL\/TLS/TLS\/SSL

Co-authored-by: Jesse Rosenberger <git@jro.cc>

EverlastingBugstopper committed:
better docs for communicating "insecure" flag tradeoffs

Co-authored-by: Jesse Rosenberger <git@jro.cc>

cecton commented on crates/rover-client/src/blocking/client.rs:
```diff
@@ -102,6 +102,10 @@ impl GraphQLClient {
+                } else if reqwest_error.is_timeout() {
+                    RoverClientError::OperationTimedOut {
+                        source: reqwest_error,
+                    }
```
That nested match-if-else could probably be reduced to a single match using the match guards. (This is just a comment, the code is fine as is.)

To reduce the bloat (independently from my previous comment) I personally usually import symbols temporarily inside the function. If the function's logic is strongly linked to backoff, it kinda makes sense to take everything in scope. Something like: `use backoff::{retry, Error::{Permanent, Transient}, ExponentialBackoff}` at the beginning of the function.

I also reduce the explicitness of the variable names when their scope gets smaller. For example, I would use very explicit and long names for global constants but I would use a single letter like `x` for `.map(|x| x...)` or `err` for a small error handling. In this case I would use `err` instead of `reqwest_error`. `backoff_strategy` => `strategy` (because the function is really all about backoff). `graphql_operation` => `func`/`operation`/`op`.

cecton replied:
```rust
Some(Duration::from_secs(if cfg!(test) { 2 } else { 10 }));
```

This needs to go in a `const`. Those `cfg!()` are evaluated at runtime and not at compile time.

cecton commented on crates/rover-client/src/error.rs:
```diff
@@ -85,6 +85,9 @@ pub enum RoverClientError {
+
     /// Encountered an error sending the request.
     #[error(transparent)]
     SendRequest(#[from] reqwest::Error),
```
Seems weird that in one case you want to use a struct variant but in the other a tuple variant despite they're both single fields... If this was private API I would say "meh whatever" but this enum is `pub` so it's best to keep the API consistent. Maybe this should go to another ticket to normalize the situation over the whole project. Feel free to create an issue based on this comment.

The `#[from]` of `SendRequest` is probably incorrect as the error wants to encapsulate a sending error but here we're are declaring that any reqwest error can be automatically converted to `SendRequest`. The `#[from]` is handy in a situation more like this:

```rust
#[derive(Error)]
pub enum Errors {
    Stuff,
    IoError(#[from] std::io::Error),
}
```
Here the intention is to encapsulate *any* I/O error so `#[from]` makes sense.

EverlastingBugstopper replied:
Yeah so I have two separate suggestions for if the request errors are related to connections or timeouts, and then a catch-all `SendRequest` error for any errors that are _not_ connection or timeout issues:

```
if reqwest_error.is_connect() {
  RoverClientError::CouldNotConnect {
    url: reqwest_error.url().cloned(),
    source: reqwest_error,
  }
} else if reqwest_error.is_timeout() {
  RoverClientError::OperationTimedOut {
    source: reqwest_error,
  }
} else {
  reqwest_error.into()
}
```

Maybe to make that more obvious I should just do `RoverClientError::SendRequest { source: reqwest_error }`

EverlastingBugstopper replied:
mm there were some other load-bearing `?` that relied on that `SendRequest` thing. anyways - i figured out a better pattern that fixes this problem and cleans up a bit of code, thanks for the :red flag: :)

cecton commented on src/utils/client.rs:
```diff
@@ -1,26 +1,70 @@
+        write!(f, "{}", self.duration.as_secs())
+    }
+}
+
```
This looks like an unnecessary wrapper. Why do you need to wrap Duration and not just use Duration directly? If it's for parsing the command line you can define a custom function for parsing

EverlastingBugstopper replied:
i like to use the trait implementations as much as possible when working with structopt - i find it easier to debug/reason about, and also lets us have all of the help text info stored in a type

cecton commented on src/cli.rs:
```diff
@@ -250,8 +259,12 @@ impl Rover {
+                    get_configured_client(
+                        self.accept_invalid_certs,
+                        self.accept_invalid_hostnames,
+                        &self.client_timeout,
```
Duration is Copy. You could make your wrapper Copy. (Or don't use a wrapper at all)

cecton commented on src/error/metadata/suggestion.rs:
```diff
@@ -141,7 +142,8 @@ impl Display for Suggestion {
             Suggestion::FixCompositionErrors => format!("The subgraph schemas you provided are incompatible with each other. See {} for more information on resolving build errors.", Cyan.normal().paint("https://www.apollographql.com/docs/federation/errors/")),
-            Suggestion::FixOperationsInSchema { graph_ref } => format!("The changes in the schema you proposed are incompatible with graph {}. See {} for more information on resolving operation check errors.", Yellow.normal().paint(graph_ref.to_string()), Cyan.normal().paint("https://www.apollographql.com/docs/studio/schema-checks/"))
+            Suggestion::FixOperationsInSchema { graph_ref } => format!("The changes in the schema you proposed are incompatible with graph {}. See {} for more information on resolving operation check errors.", Yellow.normal().paint(graph_ref.to_string()), Cyan.normal().paint("https://www.apollographql.com/docs/studio/schema-checks/")),
+            Suggestion::IncreaseClientTimeout => "You can try increasing the timeout value by passing a higher value to the --client-timeout option.".to_string()
```
To avoid unnecessary diff in the git history, you can put that final comma.

Tbh this is the first time I see that this comma is optional. I thought the rule was: for block don't use comma, for inline use comma.

o0Ignition0o commented on src/cli.rs:
```diff
@@ -250,8 +259,12 @@ impl Rover {
                 .fill(
-                    get_configured_client(self.accept_invalid_certs, self.accept_invalid_hostnames)
-                        .expect("Could not configure the request client"),
+                    get_configured_client(
```
Maybe switching to a builder might come in handy eventually

EverlastingBugstopper replied:
agreed - opened issue #844

o0Ignition0o commented on src/utils/client.rs:
```diff
@@ -1,26 +1,70 @@
+
+impl fmt::Display for ClientTimeout {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.duration.as_secs())
```
```suggestion
        write!(f, "{}s", self.duration.as_secs())
```
Could mentioning it's seconds in the display impl prove useful?

EverlastingBugstopper replied:
Yeah I had considered this actually, but the problem is that structopt uses the Display implementation to show possible values for the argument, and since we don't take them in with the `s` or `seconds` suffix, it would be a bit confusing to see that as a possible value in the `--help` output

o0Ignition0o reviewed (approved):
Kudos again on the really great documentation! 🎉

EverlastingBugstopper committed:
chore: address @cecton comments

- cleans up request errors by properly assigning codes and suggestions
- local import for backoff crate
- const the cfg!(test) check

EverlastingBugstopper committed:
chore: make ClientTimeout Copy

EverlastingBugstopper committed:
Merge branch 'main' into avery/configure-timeout

StephenBarlow commented on docs/source/configuring.md:
```diff
@@ -272,6 +272,13 @@ In some configurations (often on internal networks) users may need Rover to comm
+
+## Increase the client timeout
+
+Sometimes, you may want to make queries for large amounts of data that may take some extra time for the Studio API to process.
```
Recommendation (feel free to adjust anything I've gotten a little wrong or you disagree with):

## Increasing request timeouts

By default, Rover times out requests to the Apollo Studio API after 30 seconds. If you're executing a command that might take longer than 30 seconds to process, you can increase this timeout with the `--client-timeout` option:

```bash
rover lengthy command (whatever that would be) --client-timeout=60 # in seconds
```

StephenBarlow commented on docs/source/errors.md:
```diff
@@ -246,3 +246,9 @@ Some build errors are part of normal workflows. For instance, you may need to pu
 
+### E031
+
+This error occurs when Rover made an HTTP(S) request and it timed out.
```
"made an HTTP request that timed out." <- I'm not too stressed about needing to include the (S)

StephenBarlow commented on docs/source/errors.md:
```diff
@@ -246,3 +246,9 @@ Some build errors are part of normal workflows. For instance, you may need to pu
+
+This error occurs when Rover made an HTTP(S) request and it timed out.
+
+The client timeout that Rover sets is configurable. You can try increasing the number of seconds Rover will wait to timeout with the `--client-timeout` flag, but it's also possible that you've made a request for too much data from the Studio API, or that the Studio API is experiencing degraded performance. You can check for the latter at our [status page](https://status.apollographql.com).
```
"You can [increase Rover's request timeout](#link), but it's also possible..."

"You can check for known performance issues on our status page"

StephenBarlow replied:
That link would point to the new section in configuring.md
